### PR TITLE
[WIP; draft] Redistribute drafter kernels

### PR DESCRIPTION
**Pull request by keyboardAnt**
**Update (Aug 19, 2025):** This implementation of RDK is based on a lower bound we found for the [socp](https://github.com/sgl-project/sglang/issues/8581#issuecomment-3138384224), which we eventually found too loose. As a result, we've decided to move away from a training-free approach to an approach that requires training, as in EAGLE and FR-Spec. We'll create a new separate PR.

## Motivation

Implementing RDK from "Out-of-Vocabulary Sampling Boosts Speculative Decoding" (https://arxiv.org/abs/2506.03206).

## Benchmark

For baseline, see #8391.

```
python3 -m sglang.bench_offline_throughput --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --num-prompts 10 --speculative-algorithm EAGLE \
    --speculative-draft-model-path lmsys/sglang-EAGLE-LLaMA3-Instruct-8B --speculative-num-steps 5 \
    --speculative-eagle-topk 8 --speculative-num-draft-tokens 64 \
    --speculative-token-map thunlp/LLaMA3-Instruct-8B-FR-Spec/freq_32768.pt \
    --speculative-weaker-drafter-probs nadav-timor/meta-llama-Meta-Llama-3.1-8B-Instruct/weaker_drafter_probs.pt \
    --mem-fraction 0.7 --cuda-graph-max-bs 2 --dtype float16
```

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_worker.py [hunk 1]
```diff
@@ -905,10 +940,51 @@ def forward_draft_extend_after_decode(self, batch: ScheduleBatch):
         batch.spec_info.accept_length = accept_length_backup
         batch.return_logprob = return_logprob_backup
 
+    def _post_process_draft_logits(self, logits_output: LogitsProcessorOutput):
+        """Lower bound the ExpSumLog for out-of-vocab tokens."""
+        if not hasattr(self, "num_cold_tokens") or self.num_cold_tokens == 0:
+            return
```
[the remark this answers is not in the document]
Comment (keyboardAnt) [reply python/sglang/srt/speculative/eagle_worker.py]: Done.

File: python/sglang/srt/speculative/eagle_worker.py [hunk 2]
```diff
@@ -905,10 +940,51 @@ def forward_draft_extend_after_decode(self, batch: ScheduleBatch):
...
+        )
+
+    def _post_process_draft_probs(self, probs: torch.Tensor) -> torch.Tensor:
+        """Redistribute the probability mass of cold tokens."""
+        if (
+            not hasattr(self, "num_cold_tokens")
+            or self.num_cold_tokens == 0
+            or not hasattr(self, "weaker_drafter")
+            or self.weaker_drafter is None
+        ):
```
[the remark this answers is not in the document]
Comment (keyboardAnt) [reply python/sglang/srt/speculative/eagle_worker.py]: Done.

**Comment by keyboardAnt** [comment]
Note: for development, I'm currently ([a771699](https://github.com/sgl-project/sglang/pull/8393/commits/a771699f0fe6bf7cd10bb7be97898831bc99f797)) using the following command:
```
CUDA_LAUNCH_BLOCKING=1 python3 -m sglang.bench_offline_throughput --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --num-prompts 10 --speculative-algorithm EAGLE     --speculative-draft-model-path lmsys/sglang-EAGLE-LLaMA3-Instruct-8B --speculative-num-steps 5     --speculative-eagle-topk 8 --speculative-num-draft-tokens 64     --speculative-token-map /home/projects/dharel/nadavt/repos/sglang/python/sglang/srt/speculative/token_maps/indices_to_keep_by_target_var.pt     --speculative-weaker-drafter-probs /home/projects/dharel/nadavt/repos/sglang/python/sglang/srt/speculative/token_maps/target-probs-meta-llama-Llama-3.1-8B-Instruct-wikitext-wikitext-103-raw-v1-train.pt     --mem-fraction 0.7 --cuda-graph-max-bs 2 --dtype float16 --log-level debug  --disable-cuda-graph --enable-nan-detection
```

**Comment by keyboardAnt** [comment]
A minimal version lives in this branch: https://github.com/keyboardAnt/sglang/tree/nadav/rdk-minimal.

Below are preliminary benchmark results for different head precisions. Here, we used our `speculative-token-map` (implementing pruning by target variance).

`torch.float32`:
```
====== Offline Throughput Benchmark Result =======
Backend:                                 engine    
Successful requests:                     10        
Benchmark duration (s):                  35.40     
Total input tokens:                      1960      
Total generated tokens:                  2774      
Last generation throughput (tok/s):      22.08     
Request throughput (req/s):              0.28      
Input token throughput (tok/s):          55.37     
Output token throughput (tok/s):         78.36     
Total token throughput (tok/s):          133.73    
==================================================
```

`torch.float64`:
```
====== Offline Throughput Benchmark Result =======
Backend:                                 engine    
Successful requests:                     10        
Benchmark duration (s):                  35.81     
Total input tokens:                      1960      
Total generated tokens:                  2774      
Last generation throughput (tok/s):      21.41     
Request throughput (req/s):              0.28      
Input token throughput (tok/s):          54.74     
Output token throughput (tok/s):         77.47     
Total token throughput (tok/s):          132.21    
==================================================
```

Using `torch.float32` and the default vocab of FR-Spec (`--speculative-token-map /home/projects/dharel/nadavt/repos/sglang/python/sglang/srt/speculative/token_maps/indices_to_keep_by_tokenizer_freq.pt`):
```
====== Offline Throughput Benchmark Result =======
Backend:                                 engine    
Successful requests:                     10        
Benchmark duration (s):                  34.92     
Total input tokens:                      1960      
Total generated tokens:                  2774      
Last generation throughput (tok/s):      22.23     
Request throughput (req/s):              0.29      
Input token throughput (tok/s):          56.13     
Output token throughput (tok/s):         79.43     
Total token throughput (tok/s):          135.56    
==================================================
```

Command (same as in https://github.com/sgl-project/sglang/issues/8391):
```
CUDA_LAUNCH_BLOCKING=1 python3 -m sglang.bench_offline_throughput --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --num-prompts 10 --speculative-algorithm EAGLE \
    --speculative-draft-model-path lmsys/sglang-EAGLE-LLaMA3-Instruct-8B --speculative-num-steps 5 \
    --speculative-eagle-topk 8 --speculative-num-draft-tokens 64 \
    --speculative-token-map /home/projects/dharel/nadavt/repos/sglang/python/sglang/srt/speculative/token_maps/indices_to_keep_by_target_var.pt     --speculative-weaker-drafter-probs /home/projects/dharel/nadavt/repos/sglang/python/sglang/srt/speculative/token_maps/target-probs-meta-llama-Llama-3.1-8B-Instruct-wikitext-wikitext-103-raw-v1-train.pt \
    --mem-fraction 0.7 --cuda-graph-max-bs 2 --dtype float16 --log-level debug --disable-cuda-graph --enable-nan-detection
```

**Comment by keyboardAnt** [comment]
Benchmarking the acceptance of FR-Spec and our method (RDK), given the same vocab of 10k tokens (https://github.com/keyboardAnt/sglang/blob/nadav/acceptance/python/sglang/srt/speculative/token_maps/indices_to_keep_by_target_var.pt):
* RDK: 0.03 draft tokens
* FR-Spec: 1.33 draft tokens

⚠️ These results seem to suggest that we have an implementation bug.

---

#### Details

The code used for RDK is on branch https://github.com/keyboardAnt/sglang/tree/nadav/rdk-minimal-acceptance (commit a6367cfe6ce18063600819ad7cade7f881608b20). Command used:
```
CUDA_LAUNCH_BLOCKING=1 python3 -m sglang.bench_offline_throughput \
    --model-path meta-llama/Meta-Llama-3.1-8B-Instruct \
    --num-prompts 10 \
    --speculative-algorithm EAGLE \
    --speculative-draft-model-path lmsys/sglang-EAGLE-LLaMA3-Instruct-8B \
    --speculative-num-steps 5 --speculative-eagle-topk 8 --speculative-num-draft-tokens 64 \
    --speculative-token-map /home/projects/dharel/nadavt/repos/sglang/python/sglang/srt/speculative/token_maps/indices_to_keep_by_target_var.pt \
    --speculative-weaker-drafter-probs /home/projects/dharel/nadavt/repos/sglang/python/sglang/srt/speculative/token_maps/target-probs-meta-llama-Llama-3.1-8B-Instruct-wikitext-wikitext-103-raw-v1-train.pt \
    --mem-fraction 0.7 \
    --cuda-graph-max-bs 2 \
    --dtype float16 \
    --log-level debug \
    --disable-cuda-graph \
    --enable-nan-detection
```

The code used for FR-Spec is on branch https://github.com/keyboardAnt/sglang/tree/nadav/acceptance (commit 2e798503514ad32225fc9960afe081c694b94070). Command used: same as RDK, but without `--speculative-weaker-drafter-probs`.

**Comment by keyboardAnt** [comment]
Summary of my dev branches, initially based on top of the latest stable release at the time (`v0.4.9.post4`):

### RDK (via an approximated `socp`)

`nadav/rdk-minimal` implements a lower bound on the LogSumExp of the logits entry corresponding to the [socp](https://github.com/sgl-project/sglang/issues/8581#issuecomment-3138384224). It includes the dir `python/sglang/srt/speculative/token_maps/` with the raw token scores for target probs and variance, their corresponding token maps, and the notebook `token_maps.ipynb`.

The following branches are based on `nadav/rdk-minimal`. Experiments using these branches suggest that the aforementioned lower bound is too loose:
  * `nadav/rdk-minimal-theta`: Prints the result `socp` when using the lower bound mentioned above.
  * `nadav/rdk-minimal-acceptance`: Calculates and prints the average accepted draft tokens per request.

`nadav/rdk-minimal` implements a minimal version of the older branch `nadav/rdk`.

### Acceptance

`nadav/acceptance` implements a minimal calculation and printing of the average accepted draft tokens per request. This branch merged into the `nadav/rdk-minimal` branch to get `nadav/rdk-minimal-acceptance`.

### Oracle `socp`

`nadav/exact-socp` calculates and prints stats of the "oracle" `socp` (mean, std, min, max). The oracle `socp` is the exact ground truth value without approximation.
